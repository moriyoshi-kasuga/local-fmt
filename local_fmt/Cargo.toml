[package]
name = "local-fmt"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "This crate provides localizable format strings."
authors.workspace = true
repository.workspace = true
license.workspace = true
categories.workspace = true
keywords.workspace = true

[dependencies]
local-fmt-derive = { version = "^0", path = "../local_fmt_derive", optional = true }

[features]
default = []
derive = ["local-fmt-derive"]
selected = ["local-fmt-derive?/selected"]
global = ["local-fmt-derive?/global"]

[lib]
name = "local_fmt"

[lints]
workspace = true
