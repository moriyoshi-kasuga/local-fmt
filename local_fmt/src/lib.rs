use std::collections::HashMap;
use std::hash::Hash;

#[cfg(feature = "macros")]
pub use local_fmt_macros as macros;

pub mod message;
pub use message::*;

pub struct LocalFmt<L, M> {
    messages: HashMap<L, M>,
    global: fn() -> L,
}

impl<L: Eq + Hash, M> LocalFmt<L, M> {
    pub fn new(messages: HashMap<L, M>, global: fn() -> L) -> Self {
        Self { messages, global }
    }

    pub fn get_message(&self) -> &M {
        #[allow(clippy::expect_used)]
        self.messages
            .get(&(self.global)())
            .expect("message not found")
    }
}
