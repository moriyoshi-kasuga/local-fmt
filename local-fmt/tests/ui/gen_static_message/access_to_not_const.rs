use local_fmt::{gen_static_message, StaticMessage};

fn main() {
    let world = "world";
    const _: StaticMessage<1> = gen_static_message!("Hello! {world}");
}
