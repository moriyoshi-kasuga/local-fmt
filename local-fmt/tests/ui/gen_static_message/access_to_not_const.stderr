error[E0435]: attempt to use a non-constant value in a constant
 --> tests/ui/gen_static_message/access_to_not_const.rs:5:33
  |
5 |     const _: StaticMessage<1> = gen_static_message!("Hello! {world}");
  |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ non-constant value
  |
  = note: this error originates in the macro `gen_static_message` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider using `let` instead of `const`
  |
5 |     let _: StaticMessage<1> = gen_static_message!("Hello! {world}");
  |     ~~~

error[E0308]: mismatched types
 --> tests/ui/gen_static_message/access_to_not_const.rs:5:33
  |
5 |     const _: StaticMessage<1> = gen_static_message!("Hello! {world}");
  |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `1`, found `0`
  |
  = note: expected struct `RefMessage<'static, 1>`
             found struct `RefMessage<'_, 0>`
  = note: this error originates in the macro `gen_static_message` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unused variable: `world`
 --> tests/ui/gen_static_message/access_to_not_const.rs:4:9
  |
4 |     let world = "world";
  |         ^^^^^ help: if this is intentional, prefix it with an underscore: `_world`
  |
  = note: `#[warn(unused_variables)]` on by default
